frontend/src/metabase/components/Triggerable/Triggerable.tsx

```diff
@@ -1,42 +1,86 @@
-/* eslint-disable react/prop-types */
-import React, { Component } from "react";
-
-import cx from "classnames";
 import styled from "@emotion/styled";
+import cx from "classnames";
+import React, {
+  CSSProperties,
+  Component,
+  ComponentType,
+  ElementType,
+  ReactNode,
+  SyntheticEvent,
+  createRef,
+} from "react";
+
 import { isObscured } from "metabase/lib/dom";
 
-import Tooltip from "metabase/core/components/Tooltip";
+import { Children } from "./Children";
+import { TriggerElement } from "./TriggerElement";
+import { RenderProp, RenderTriggerElement } from "./types";
 
 const Trigger = styled.a``;
 
+type TriggerableComponentProps<P extends Record<string, unknown>> = P & {
+  isOpen: boolean;
+  sizeToFit?: boolean;
+  target?: () => EventTarget | null;
+  onClose?: (event: SyntheticEvent) => void;
+};
+
+interface Props {
+  as?: ElementType | ComponentType;
+  children:
+    | ReactNode
+    | RenderProp<{ onClose: (event: SyntheticEvent) => void }>;
+  closeOnObscuredTrigger?: boolean;
+  disabled?: boolean;
+  isInitiallyOpen?: boolean;
+  isOpen?: boolean;
+  target?: () => EventTarget | null;
+  triggerClasses?: string;
+  triggerClassesClose?: string;
+  triggerClassesOpen?: string;
+  triggerElement?: ReactNode | RenderTriggerElement;
+  triggerId?: string;
+  triggerStyle?: CSSProperties;
+  onClose?: (event: SyntheticEvent) => void;
+}
+
+interface State {
+  isOpen: boolean;
+}
+
 // higher order component that takes a component which takes props "isOpen" and optionally "onClose"
 // and returns a component that renders a <a> element "trigger", and tracks whether that component is open or not
+const _Triggerable = <P extends Record<string, unknown>>(
```

Comment:
Could we use a generic type here instead of `Record`? E.g. `P extends Something`, where `Something` would have required props if we have them

Reply:
I'm not sure if I follow - is this what you have in mind?

```suggestion
const _Triggerable = <R extends Record<string, unknown>, P extends R>(
```